### DISCUSSION TITLE
Requesting more information around rabbit_mqtt_retained_msg_store

### DISCUSSION THREAD

**Discussion by jonesjaycob** [rabbitmq/rabbitmq-server#8787] [Other]
I'm looking to use RabbitMQ MQTT with a database option such as Cassandra. The documentation states this is possible, but I cannot find anything else to make this option feasible. Has anyone successfully implemented Cassandra as the retained message store? Or can someone supply more information regarding this? 

From the RabbitMQ [Docs](https://www.rabbitmq.com/mqtt.html#retained):

_These implementations are suitable for development but sometimes won't be for production needs. MQTT 3.1 specification does not define consistency or replication requirements for retained message stores, therefore RabbitMQ allows for custom ones to meet the consistency and availability needs of a particular environment. For example, stores based on [Riak](http://basho.com/riak/) and [Cassandra](http://cassandra.apache.org/) would be suitable for most production environments as those data stores provide [tunable consistency](https://github.com/basho/basho_docs/blob/master/content/riak/kv/2.2.3/using/reference/strong-consistency.md)._

_Message stores must implement the rabbit_mqtt_retained_msg_store behaviour._

**Comment by michaelklishin** [comment]
[`rabbit_mqtt_retained_msg_store`](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store.erl) is a behavior (an interface) that anyone can implement (like here for [DETS](https://github.com/rabbitmq/rabbitmq-server/blob/main/deps/rabbitmq_mqtt/src/rabbit_mqtt_retained_msg_store_dets.erl), a node local store that ships with Erlang/OTP), then ship in a plugin, enable it, and specify the module as
`rabbitmq_mqtt.retained_message_store` in the `advanced.config`.

Or you can look into contributing a solution for https://github.com/rabbitmq/rabbitmq-server/issues/8824 and similar issues. There are no promises when the core team might get to it,
could easily be in the second half of 2024. But this is open source software.

**Comment by lukebakken** [comment]
https://github.com/cqerl/cqerl
